sampling_fo2/parser/mln_grammar.py

```diff
@@ -0,0 +1,26 @@
+from .fol_grammar import function_free_logic_grammar
+
+domain_grammar = r"""
+    domain: domain_name "=" domain_spec
+    domain_name: CNAME
+    ?domain_spec: INT               -> int_domain
+        | ("{" domain_elements "}") -> set_domain
+    domain_elements: element ("," element)*
+    element: CNAME
+"""
+
+cardinality_grammar = r"""
+    cardinalities: cardinality*
+    cardinality: "|" predicate "|" comparator cardinality_param
+    cardinality_param: INT
+"""
+
+rule_grammar = r"""
+    rules: rule*
+    rule: weighting? ffl "."
```

Comment:
The period "." means the rule is a hard rule. Soft rules don't have this symbol.